pkg/daemon/ceph/osd/remove.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package osd
+
+import (
+	"fmt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+// RemoveOSDs purges a list of OSDs from the cluster
+func RemoveOSDs(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdsToRemove []string) error {
+
+	// Generate the ceph config for running ceph commands similar to the operator
+	if err := writeCephConfig(context, clusterInfo); err != nil {
+		return errors.Wrap(err, "failed to write the ceph config")
+	}
+
+	// Check if the OSD is down
+	osdDump, err := client.GetOSDDump(context, clusterInfo)
+	if err != nil {
+		return errors.Wrap(err, "failed to get osd dump")
+	}
+
+	for _, osdStatus := range osdDump.OSDs {
+		id64, err := osdStatus.OSD.Int64()
+		if err != nil {
+			continue
+		}
+		id := int(id64)
+
+		logger.Debugf("validating status of osd.%d", id)
+
+		status, _, err := osdDump.StatusByID(int64(id))
+		if err != nil {
+			return err
+		}
+
+		const upStatus int64 = 1
+		if status == upStatus {
+			logger.Debugf("osd.%d is healthy.It cannot be removed unless it is 'down'", id)
+			continue
+		}
+		logger.Debugf("osd.%d is marked 'DOWN'", id)
+	}
+
+	for _, osdIDStr := range osdsToRemove {
+		logger.Infof("removing OSD %q", osdIDStr)
+		osdID, err := strconv.Atoi(osdIDStr)
+		if err != nil {
+			logger.Errorf("invalid OSD ID: %s. %v", osdIDStr, err)
+			continue
+		}
+
+		if err := removeOSD(context, clusterInfo, osdID); err != nil {
+			logger.Errorf("failed to remove OSD %d. %v", osdID, err)
+		}
+	}
+
+	return nil
+}
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) error {
```

Comment:
When this func is called, the decision should already be made that we are going to do everything to clean up the OSD. Removing all the resources is a best effort, meaning that we should try to continue removing the OSD even if some of the resources aren't found. If resources are not found, we just log it and continue with the next command.